File: plateform/backend/app/schemas/detections.py
from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.incidents import Classification, IncidentSummary, SeverityTier


class DetectionStatus(str, Enum):
    new = "new"
    queued = "queued"
    under_review = "under_review"
    dismissed = "dismissed"
    promoted = "promoted"
    merged = "merged"


class DetectionPriority(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"


class DetectionSortField(str, Enum):
    detected_at = "detected_at"
    priority = "priority"
    score = "score"
    status = "status"


class SortOrder(str, Enum):
    asc = "asc"
    desc = "desc"


class DetectionEvidenceItem(BaseModel):
    metric: str
    observed_value: float
    baseline_value: float | None = None
    threshold_value: float | None = None
    unit: str
    evidence_type: str
    reading_start_time: datetime
    reading_end_time: datetime
    details: dict[str, Any] = Field(default_factory=dict)


class DetectionSummary(BaseModel):
    id: str
    detection_number: str
    status: DetectionStatus
    priority: DetectionPriority
    anomaly_score: float = Field(ge=0, le=1)
    trigger_reason: str
    reason_codes: list[str]
    rule_code: str
    rule_name: str
    rule_version: int
    sensor_id: str
    sensor_name: str
    zone: str
    pipeline_segment: str | None = None
    detected_at: datetime
    window_start: datetime
    window_end: datetime
    data_mode: str
    incident_id: str | None = None
    reviewed_by: str | None = None
    review_started_at: datetime | None = None
    dismissed_at: datetime | None = None
    dismissal_reason: str | None = None
    merged_into_detection_id: str | None = None
    resolved_at: datetime | None = None
    updated_at: datetime | None = None
    allowed_actions: list[str] = Field(default_factory=list)
    screening_priority_label: str = "Screening priority"
    awaiting_agent_investigation: bool = False
    has_agent_finding: bool = False


class DetectionListResponse(BaseModel):
    items: list[DetectionSummary]
    total: int = Field(ge=0)
    data_mode: str = "simulated"


class DetectionDetail(BaseModel):
    id: str
    detection_number: str
    status: DetectionStatus
    priority: DetectionPriority
    anomaly_score: float = Field(ge=0, le=1)
    trigger_reason: str
    reason_codes: list[str]
    rule_code: str
    rule_name: str
    rule_version: int
    sensor_id: str
    sensor_name: str
    asset_status: str
    zone: str
    pipeline_segment: str | None = None
    pipeline_segment_id: str | None = None
    criticality: int | None = None
    population_served: int | None = None
    detected_at: datetime
    window_start: datetime
    window_end: datetime
    reading_count: int
    correlation_key: str
    score_explanation: dict[str, Any] = Field(default_factory=dict)
    evidence_summary: dict[str, Any] = Field(default_factory=dict)
    evidence: list[DetectionEvidenceItem] = Field(default_factory=list)
    related_detections: list[DetectionSummary] = Field(default_factory=list)
    incident_id: str | None = None
    telemetry_freshness: str | None = None
    latest_reading_at: datetime | None = None
    data_mode: str = "simulated"
    reviewed_by: str | None = None
    review_started_at: datetime | None = None
    dismissed_at: datetime | None = None
    dismissal_reason: str | None = None
    merged_into_detection_id: str | None = None
    resolved_at: datetime | None = None
    updated_at: datetime | None = None
    allowed_actions: list[str] = Field(default_factory=list)
    screening_priority_label: str = "Screening priority"
    awaiting_agent_investigation: bool = False
    has_agent_finding: bool = False
    agent_finding: dict[str, Any] | None = None


class DetectionEvidenceResponse(BaseModel):
    detection_id: str
    items: list[DetectionEvidenceItem]
    data_mode: str = "simulated"


class DetectionQueueStats(BaseModel):
    new: int = Field(ge=0)
    queued: int = Field(ge=0)
    under_review: int = Field(ge=0)
    high_priority: int = Field(ge=0)
    critical_priority: int = Field(ge=0)
    total: int = Field(ge=0)
    last_run_at: datetime | None = None
    last_run_deduplicated: int | None = None
    highest_priority: DetectionSummary | None = None
    data_mode: str = "simulated"


class TelemetryWindow(BaseModel):
    start: datetime
    end: datetime


class AgentSensorIdentity(BaseModel):
    id: str
    name: str
    operational_status: str
    manufacturer: str | None = None
    model: str | None = None


class AgentNetworkMetadata(BaseModel):
    zone: str
    signal_strength_dbm: int | None = None
    packet_loss_pct: float | None = None
    telemetry_freshness: str | None = None
    latest_reading_at: datetime | None = None


class AgentRuleTrigger(BaseModel):
    rule_code: str
    rule_name: str
    rule_version: int
    reason_codes: list[str]
    trigger_reason: str
    threshold: float | None = None
    secondary_threshold: float | None = None


class AgentPipelineContext(BaseModel):
    segment_id: str | None = None
    segment_name: str | None = None
    criticality: int | None = None
    population_served: int | None = None


class InvestigationAgentInputV1(BaseModel):
    """Future Investigation Agent input. No agent is invoked when this is built."""

    schema_version: Literal["1"] = "1"
    detection_id: str
    detected_at: datetime
    telemetry_window: TelemetryWindow
    recent_readings_reference: dict[str, Any]
    sensor: AgentSensorIdentity
    network_metadata: AgentNetworkMetadata
    rule_triggers: list[AgentRuleTrigger]
    structured_evidence: list[DetectionEvidenceItem]
    pipeline_context: AgentPipelineContext
    criticality: int | None = None
    population_served: int | None = None
    current_asset_status: str
    data_mode: str = "simulated"
    agent_executed: Literal[False] = False
    note: str = (
        "No investigation agent has run. This payload is the versioned input contract "
        "for a future agent that will classify leak, connectivity, sensor fault, or normal."
    )


class DismissalReason(str, Enum):
    false_positive = "false_positive"
    sensor_fault = "sensor_fault"
    planned_operation = "planned_operation"
    duplicate = "duplicate"
    insufficient_evidence = "insufficient_evidence"
    other = "other"


class InvestigationEventType(str, Enum):
    review_started = "review_started"
    note_added = "note_added"
    dismissed = "dismissed"
    reopened = "reopened"
    merged = "merged"
    promoted = "promoted"


def _trim_actor(value: str) -> str:
    """Temporary development identity. Not trusted authentication."""
    trimmed = value.strip()
    if not trimmed:
        raise ValueError("actor_name is required")
    return trimmed


class ActorRequest(BaseModel):
    """Temporary development identity. Not trusted authentication. Not a user session."""

    actor_name: str = Field(
        ...,
        min_length=1,
        max_length=120,
        description="Temporary development identity. Not trusted authentication.",
    )

    @field_validator("actor_name")
    @classmethod
    def validate_actor_name(cls, value: str) -> str:
        return _trim_actor(value)


class StartReviewRequest(ActorRequest):
    note: str | None = Field(default=None, max_length=2000)

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class AddNoteRequest(ActorRequest):
    note: str = Field(..., min_length=1, max_length=2000)

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("note is required")
        return trimmed


class DismissDetectionRequest(ActorRequest):
    reason_code: DismissalReason
    note: str | None = Field(default=None, max_length=2000)

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class ReopenDetectionRequest(ActorRequest):
    note: str | None = Field(default=None, max_length=2000)

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class MergeDetectionRequest(ActorRequest):
    target_detection_id: str = Field(..., min_length=1, max_length=32)
    note: str | None = Field(default=None, max_length=2000)

    @field_validator("target_detection_id")
    @classmethod
    def validate_target(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("target_detection_id is required")
        return trimmed.upper()

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class PromoteDetectionRequest(ActorRequest):
    title: str = Field(..., min_length=1, max_length=300)
    severity: SeverityTier
    classification: Classification
    summary: str = Field(..., min_length=1, max_length=2000)
    note: str | None = Field(default=None, max_length=2000)

    @field_validator("title", "summary")
    @classmethod
    def validate_required_text(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("this field is required")
        return trimmed

    @field_validator("note")
    @classmethod
    def validate_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class InvestigationEventItem(BaseModel):
    id: str
    event_type: InvestigationEventType
    from_status: DetectionStatus | None = None
    to_status: DetectionStatus | None = None
    actor_name: str
    note: str | None = None
    reason_code: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    target_detection_id: str | None = None
    incident_id: str | None = None


class InvestigationHistoryResponse(BaseModel):
    detection_id: str
    events: list[InvestigationEventItem]


class PromoteDetectionResponse(BaseModel):
    detection: DetectionDetail
    incident: IncidentSummary

File: plateform/frontend/src/types/detections.ts
export type DetectionStatus =
  | "new"
  | "queued"
  | "under_review"
  | "dismissed"
  | "promoted"
  | "merged";

export type DetectionPriority = "low" | "medium" | "high" | "critical";

export type DetectionSortField = "detected_at" | "priority" | "score" | "status";

export type SortOrder = "asc" | "desc";

export interface DetectionFilters {
  search: string;
  status: DetectionStatus | "";
  priority: DetectionPriority | "";
  rule: string;
  zone: string;
  sensor: string;
  start: string;
  end: string;
  sort_by: DetectionSortField;
  sort_order: SortOrder;
}

export interface DetectionSummary {
  id: string;
  detection_number: string;
  status: DetectionStatus;
  priority: DetectionPriority;
  anomaly_score: number;
  trigger_reason: string;
  reason_codes: string[];
  rule_code: string;
  rule_name: string;
  rule_version: number;
  sensor_id: string;
  sensor_name: string;
  zone: string;
  pipeline_segment: string | null;
  detected_at: string;
  window_start: string;
  window_end: string;
  data_mode: string;
  incident_id: string | null;
  reviewed_by: string | null;
  review_started_at: string | null;
  dismissed_at: string | null;
  dismissal_reason: string | null;
  merged_into_detection_id: string | null;
  resolved_at: string | null;
  updated_at: string | null;
  allowed_actions: string[];
  screening_priority_label: string;
  awaiting_agent_investigation: boolean;
  has_agent_finding: boolean;
}

export interface DetectionEvidenceItem {
  metric: string;
  observed_value: number;
  baseline_value: number | null;
  threshold_value: number | null;
  unit: string;
  evidence_type: string;
  reading_start_time: string;
  reading_end_time: string;
  details: Record<string, unknown>;
}

export interface DetectionDetail extends DetectionSummary {
  asset_status: string;
  pipeline_segment_id: string | null;
  criticality: number | null;
  population_served: number | null;
  reading_count: number;
  correlation_key: string;
  score_explanation: {
    weights?: Record<string, number>;
    factors?: Record<string, number>;
    weighted?: Record<string, number>;
    anomaly_score?: number;
  };
  evidence_summary: Record<string, unknown>;
  evidence: DetectionEvidenceItem[];
  related_detections: DetectionSummary[];
  telemetry_freshness: string | null;
  latest_reading_at: string | null;
  agent_finding: {
    classification: string;
    severity_tier: number;
    confidence_score: number;
    anomaly_score?: number | null;
    operator_justification: string | null;
    mapping_status: string;
    run_id: string | null;
    external_cluster_id: string | null;
  } | null;
}

export interface DetectionListResponse {
  items: DetectionSummary[];
  total: number;
  data_mode: string;
}

export interface DetectionQueueStats {
  new: number;
  queued: number;
  under_review: number;
  high_priority: number;
  critical_priority: number;
  total: number;
  last_run_at: string | null;
  last_run_deduplicated: number | null;
  highest_priority: DetectionSummary | null;
  data_mode: string;
}

export interface InvestigationAgentInputV1 {
  schema_version: "1";
  detection_id: string;
  detected_at: string;
  telemetry_window: { start: string; end: string };
  recent_readings_reference: Record<string, unknown>;
  sensor: {
    id: string;
    name: string;
    operational_status: string;
    manufacturer: string | null;
    model: string | null;
  };
  network_metadata: {
    zone: string;
    signal_strength_dbm: number | null;
    packet_loss_pct: number | null;
    telemetry_freshness: string | null;
    latest_reading_at: string | null;
  };
  rule_triggers: Array<{
    rule_code: string;
    rule_name: string;
    rule_version: number;
    reason_codes: string[];
    trigger_reason: string;
    threshold: number | null;
    secondary_threshold: number | null;
  }>;
  structured_evidence: DetectionEvidenceItem[];
  pipeline_context: {
    segment_id: string | null;
    segment_name: string | null;
    criticality: number | null;
    population_served: number | null;
  };
  criticality: number | null;
  population_served: number | null;
  current_asset_status: string;
  data_mode: string;
  agent_executed: false;
  note: string;
}

export type DismissalReason =
  | "false_positive"
  | "sensor_fault"
  | "planned_operation"
  | "duplicate"
  | "insufficient_evidence"
  | "other";

export type InvestigationEventType =
  | "review_started"
  | "note_added"
  | "dismissed"
  | "reopened"
  | "merged"
  | "promoted";

export type DetectionAction = "start_review" | "add_note" | "dismiss" | "merge" | "promote" | "reopen";

export interface InvestigationEventItem {
  id: string;
  event_type: InvestigationEventType;
  from_status: DetectionStatus | null;
  to_status: DetectionStatus | null;
  actor_name: string;
  note: string | null;
  reason_code: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  target_detection_id: string | null;
  incident_id: string | null;
}

export interface InvestigationHistoryResponse {
  detection_id: string;
  events: InvestigationEventItem[];
}

export interface ActorPayload {
  actor_name: string;
}

export interface StartReviewPayload extends ActorPayload {
  note?: string;
}

export interface AddNotePayload extends ActorPayload {
  note: string;
}

export interface DismissPayload extends ActorPayload {
  reason_code: DismissalReason;
  note?: string;
}

export interface ReopenPayload extends ActorPayload {
  note?: string;
}

export interface MergePayload extends ActorPayload {
  target_detection_id: string;
  note?: string;
}

export interface PromotePayload extends ActorPayload {
  title: string;
  severity: "tier_1" | "tier_2" | "tier_3";
  classification:
    | "confirmed_leak"
    | "suspected_leak"
    | "connectivity_degradation"
    | "sensor_fault"
    | "pressure_anomaly"
    | "normal_demand_spike"
    | "insufficient_data";
  summary: string;
  note?: string;
}

export interface PromoteResponse {
  detection: DetectionDetail;
  incident: {
    id: string;
    incident_number: string;
    title: string;
    classification: PromotePayload["classification"];
    severity: PromotePayload["severity"];
    status: string;
  };
}

export interface ApiErrorDetail {
  message: string;
  code?: string;
  incident_id?: string;
  target_detection_id?: string;
}
